lib/backend/fs/impl.go

```diff
@@ -0,0 +1,310 @@
+	filename := path.Join(dirPath, key)
+	f, err := os.OpenFile(filename, os.O_WRONLY|os.O_CREATE|os.O_EXCL, defaultFileMode)
+	if err != nil {
+		if os.IsExist(err) {
+			return trace.AlreadyExists("%s/%s already exists", dirPath, key)
+		}
+		return trace.ConvertSystemError(err)
+	}
+	defer f.Close()
+	n, err := f.Write(val)
+	if err == nil && n < len(val) {
+		return trace.Wrap(io.ErrShortWrite)
+	}
+	bk.applyTTL(dirPath, key, ttl)
+	return nil
+}
+
+// UpsertVal updates or inserts value with a given TTL into a bucket
+// ForeverTTL for no TTL
+func (bk *Backend) UpsertVal(bucket []string, key string, val []byte, ttl time.Duration) error {
+	log.Debugf("fs.UpsertVal(%s/%s) '%v'", strings.Join(bucket, "/"), key, string(val))
+	// create the directory:
+	dirPath := path.Join(bk.RootDir, path.Join(bucket...))
+	err := os.MkdirAll(dirPath, defaultDirMode)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	// create the (or overwrite existing) file (AKA "key"):
+	return ioutil.WriteFile(path.Join(dirPath, key), val, defaultFileMode)
+}
+
+// GetVal return a value for a given key in the bucket
+func (bk *Backend) GetVal(bucket []string, key string) ([]byte, error) {
+	log.Debugf("fs.GetVal(%s/%s)", strings.Join(bucket, "/"), key)
+	dirPath := path.Join(path.Join(bk.RootDir, path.Join(bucket...)))
+	expired, err := bk.checkTTL(dirPath, key)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	if expired {
+		bk.DeleteKey(bucket, key)
+		return nil, trace.NotFound("key '%s' is not found", key)
+	}
+	return ioutil.ReadFile(path.Join(dirPath, key))
+}
+
+// DeleteKey deletes a key in a bucket
+func (bk *Backend) DeleteKey(bucket []string, key string) error {
+	dirPath := path.Join(bk.RootDir, path.Join(bucket...))
+	if err := os.Remove(bk.ttlFile(dirPath, key)); err != nil {
+		if !os.IsNotExist(err) {
+			log.Warn(err)
+		}
+	}
+	return trace.ConvertSystemError(os.Remove(
+		path.Join(dirPath, key)))
+}
+
+// DeleteBucket deletes the bucket by a given path
+func (bk *Backend) DeleteBucket(parent []string, bucket string) error {
+	return trace.ConvertSystemError(os.RemoveAll(
+		path.Join(path.Join(bk.RootDir, path.Join(parent...)), bucket)))
+}
+
+// AcquireLock grabs a lock that will be released automatically in TTL
+func (bk *Backend) AcquireLock(token string, ttl time.Duration) (err error) {
+	log.Debugf("fs.AcquireLock(%s)", token)
+	lockPath := path.Join(bk.LocksDir, token)
+	var f *os.File
+	for {
+		f, err = os.OpenFile(lockPath, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0600)
+		if err == nil { // success
+			defer f.Close()
+			break
+		}
+		if os.IsExist(err) { // locked? wait and repeat:
+			bk.Clock.Sleep(time.Millisecond * 100)
+			continue
+		}
+		return trace.ConvertSystemError(err)
+	}
+
+	// protect the locks map:
+	bk.Lock()
+	defer bk.Unlock()
+	bk.locks[token] = f.Name()
+
+	// start the goroutine which will release lock after the given TTL
+	if ttl != backend.Forever {
+		go func() {
+			bk.Clock.Sleep(ttl)
+			bk.ReleaseLock(token)
```

Comment:
it is possible that process crashes here and the lock will stay forever. To prevent this I suggest to use:

https://github.com/coreos/rkt/blob/master/pkg/lock/file.go